### PR TITLE
Add tools for Route optimization (TSP) and enhance tool use functionality

### PR DESCRIPTION
**Pull request by himloul**
- Primarily, added tools for route optimization, specifically a simple heuristic for solving the Traveling Salesman Problem (TSP). Used the Google Maps Platform Python library.
- Updated the Assistant class to efficiently handle the tool definitions and Pydantic classes using the `@toolRegistry` decorator.
- Additionally, the Assistant now enables multi-step and parallel tool use.
- Related to the Streamlit app, changed to FastAPI for testing. @tomcotter7 Let me know if this change is convenient.

### CODE REVIEW

**Comment by tomcotter7** [comment]
@himloul  - do you want me to review or wait until it's not a draft?

**Comment by himloul** [comment]
@tomcotter7 The pull request is ready for review now. I had to update the API path for route-related queries

**Comment by tomcotter7** [comment]
No problem! I will review in the morning :)

File: swag/assistant.py [hunk 1]
```diff
@@ -1,32 +1,137 @@
+from typing import List, Dict, Any, Optional
+from pathlib import Path
 from anthropic import Anthropic
-from anthropic.types import Message
-import os
-from dotenv import load_dotenv
-from typing import Any
+from anthropic.types import TextBlock, ToolUseBlock
+import base64
+from swags.tools import ToolRegistry
```
Comment (tomcotter7) [inline swag/assistant.py]: I think this should be `swag.tools`

File: swag/assistant.py [hunk 2]
```diff
@@ -1,32 +1,137 @@
...
 
-    client = Anthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))
-    response = client.messages.create(
-            model="claude-3-5-haiku-latest",
-            max_tokens=1024,
-            system=system,
-            messages=messages,
-            tools=tools
-    )
+    def define_tools(self, tools: List[type]):
```
Comment (tomcotter7) [inline swag/assistant.py]: should be `List[Any]` or `List[Dict]`

File: swag/assistant.py [hunk 3]
```diff
@@ -1,32 +1,137 @@
...
+                model=self.model,
+                max_tokens=self.max_tokens,
+                tools=self.tools or [],
+                system=self.system or "",
+                messages=self.messages,
+            )
+
+            assistant_message = {"role": "assistant", "content": []}
+            for content in response.content:
+                if isinstance(content, TextBlock):
```
Comment (tomcotter7) [inline swag/assistant.py:90]: We need to yield the response back so we can stream it.

File: swag/tools.py [hunk 4]
```diff
@@ -41,59 +62,170 @@ def validate_types(self) -> Self:
...
+    """Get a matrix of distances between multiple origins and destinations."""
+
+    origins: list[tuple[float, float]] = Field(
+        description="List of origin locations as tuples of (latitude, longitude)"
+    )
+    destinations: list[tuple[float, float]] = Field(
+        description="List of destination locations as tuples of (latitude, longitude)"
+    )
+    mode: str = Field(
+        description="The mode of transportation (driving, walking, bicycling, transit)",
```
Comment (tomcotter7) [inline swag/tools.py:85]: I think some potential options should be added here to quickly fail if the model get it's wrong.

File: swag/assistant.py (hunk 1 above)
Comment (himloul) [reply swag/assistant.py]: Right, corrected

File: swag/assistant.py (hunk 2 above)
Comment (himloul) [reply swag/assistant.py]: Resolved

File: swag/assistant.py (hunk 3 above)
Comment (himloul) [reply swag/assistant.py:90]: Updated. when invoking the endpoint, the blocks content (object) are streamed

File: swag/tools.py (hunk 4 above)
Comment (himloul) [reply swag/tools.py:85]: Right, updated

**Comment by tomcotter7** [comment]
changes look good @himloul - can we squash & merge?

**Comment by himloul** [comment]
Good, if it looks good to you, then yes

**Pull request merged**